### PR TITLE
Add update-rules command to update publishing-bot rules

### LINKED ISSUE #252: Create a tool to update publishing-bot rules for new releases
See https://github.com/kubernetes/sig-release/pull/1504 for more context.

The publishing-bot copies/publishes code from [staging directories](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io) to their [own repos](https://github.com/search?q=topic%3Ak8s-staging+org%3Akubernetes&type=Repositories). The configuration for how the branches should be published are defined in the `master` branch of the main k/k repo: https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/rules.yaml

For example, in this code snippet - https://github.com/kubernetes/kubernetes/blob/6572fe4d90173a1e97fda3e0cc28acc95e3f560a/staging/publishing/rules.yaml#L197-L209:

- The code in the `release-1.21` branch of the [client-go staging directory](https://github.com/kubernetes/kubernetes/tree/release-1.21/staging/src/k8s.io/client-go) is published to the [`release-1.21` branch](https://github.com/kubernetes/client-go/tree/release-1.21) of the client-go repo.
- The go.mod for the `release-1.21` branch of the client-go repo uses the `release-1.21` of api and apimachinery repos as dependencies.

This issue involves creating a new tool in Go (can go in the cmd/ directory) to update rules automatically. For example, the tool should automatically update rules.yaml in k/k to add the new branches for `release-1.21` like in https://github.com/kubernetes/kubernetes/pull/100616.

Adding new branches would basically involve copying over the `master` branch contents and updating the branch name. Additionally, the `go: <version>` field can be added to new branches too.

### PR DESCRIPTION
**Pull request by navidshaikh**
### Description:
 - Add `cmd/update-rules` to update publishing-bot rules
 - Required flags: `--rules`, `--branch`
 - Exports the updated rules to file pointed by `-o` otherwise on prints stdout
 - Pin go version provided by `--go` flag else leave the field empty (`omitempty`)
 - Refer master branch rule per destination repo to add/update rule for provided branch

### Usage:
```
Usage:  update-rules --branch BRANCH --rules PATH [--go VERSION | -o PATH]
```
 
### Examples:
  - Update rules for branch release-1.21 with go version 1.16.4
  ```
    update-rules -branch release-1.21 -go 1.16.4 -rules /go/src/k8s.io/kubernetes/staging/publishing/rules.yaml
   ```
  - Update rules using URL to input rules file
  ```
  update-rules -branch release-1.21 -go 1.16.4 -rules https://raw.githubusercontent.com/kubernetes/kubernetes/master/staging/publishing/rules.yaml
  ```

  - Update rules and export to /tmp/rules.yaml
```
  update-rules -branch release-1.22 -go 1.17.1 -o /tmp/rules.yaml -rules /go/src/k8s.io/kubernetes/staging/publishing/rules.yaml
```

fixes #252 

TODO:
- [x] command
- [x] tests
- [x] docs

### CODE REVIEW

File: cmd/publishing-bot/config/rules.go
Comment (nikhita) [inline cmd/publishing-bot/config/rules.go:34]: Why is this omitempty?

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:40]: nit: since publishing-bot is used by projects other than kubernetes, we try not to introduce any k8s-specific terms in the code. would be good to remove "k/k" here :)

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:59]: (applies everywhere) s/rule/rules to match the "rules.yaml" nomenclature

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:86]: nit:

```suggestion
		glog.Fatalf("cannot load rules file %q: %v", *rulesFile, err)
```

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:50]:
```suggestion
  update-rules -branch release-1.22 -go 1.16.3 -o /tmp/updated_rules.yaml -rule /go/src/k8s.io/kubernetes/staging/publishing/rules.yaml`
```

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:70]: Since this is not specific to k/k, we can probably skip branch validation

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:112]: We should skip update for the repo and log a warning in this case instead of failing because there can be some staging directories that have been removed in master, but we'll need to continue publishing the old branches.

Example: node-api - https://github.com/kubernetes/kubernetes/blob/f137c4777095b3972e2dd71a01365d47be459389/staging/publishing/rules.yaml#L1908-L1924

File: cmd/publishing-bot/config/rules.go
Comment (nikhita) [reply cmd/publishing-bot/config/rules.go:34]: Running the script generates a diff with these fields, so would be good to add omitempty for them:
```
skip-source-branches: []
skip-godeps: false
skip-gomod: false
skip-tags: false
```

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:116]: We'll also need to update the branch for dependencies

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:120]: I think we can unconditionally update the go version here, without needed to check if a go version was specified for the master branch

File: cmd/update-rules/main.go
Comment (nikhita) [inline cmd/update-rules/main.go:146]: This removes comments...maybe consider using go-yaml/yaml to preserve comments (https://github.com/go-yaml/yaml/issues/132)

**Review by nikhita — COMMENTED** [review]
@navidshaikh thanks for working on this, left some suggestions :)

File: cmd/update-rules/main.go
Comment (navidshaikh) [reply cmd/update-rules/main.go]: changed to only update the go version if specified via CLI flag, else leave untouched / omitempty

File: cmd/update-rules/main.go
Comment (navidshaikh) [reply cmd/update-rules/main.go:136]: The comments are removed while the yaml content are deserialized into native ds. So at this point we don't have comments at all. To preserve comments, we'd need to load them into `yaml.Node{}` and navigate using it (we may also need to update the way rules loading and validations are done in `cmd/publishing-bot/config/rules.go`). It's available in yaml.v3, currently we are using yaml.v2 here.

**Comment by navidshaikh** [comment]
Updated PR to print the updated rules on stdout if no `-o` flag is specified, this allows providing input rules by URL (using `--rules` flag).

**Comment by navidshaikh** [comment]
/retitle Add update-rules command to update publishing-bot rules

**Comment by navidshaikh** [comment]
Rebased and squashed the commits.

**Comment by navidshaikh** [comment]
/cc @nikhita @cpanato 

this is ready for review, PTAL at PR desc for usage/examples

**Comment by palnabarun** [comment]
I had tried to generate the rules in https://github.com/kubernetes/kubernetes/pull/103844 using the code here.

Some general observations:
1. The comments in `rules.yaml` get removed after running this. I think this is a blocker to use this tool.
2. The order of the keys in dictionaries gets changed. This is not a blocker as the key reordering happens only once. Post the initial change, the diff would be lower as the keys would be sorted in a previous rule update.
3. We should have a Makefile with a recipe to install the `updates-rules` binary. This is also good to have as other parties can install the tool consistently.

**Comment by xmudrii** [comment]
@navidshaikh @nikhita @palnabarun Is there any update on this PR? It would be nice to try to get this done -- it's definitely a very useful tool.

/remove-lifecycle stale

**Comment by justaugustus** [comment]
@navidshaikh -- Would you mind rebasing this?

**Comment by navidshaikh** [comment]
> Is there any update on this PR? It would be nice to try to get this done -- it's definitely a very useful tool.

@xmudrii sorry for late reply! will work through the feedback now.


@palnabarun : thanks for the review!

> The comments in rules.yaml get removed after running this. I think this is a blocker to use this tool.

This tool uses the existing utils (using yaml.v2) to load the rules. To retain the comments I think we'll need yaml.v3. 

Question: Should we have alternate yaml loading utils (with yaml.v3) and use it with this tool to retain the comments?
 or move the [existing utils](https://github.com/kubernetes/publishing-bot/blob/master/cmd/publishing-bot/config/rules.go#L87) to yaml.v3?

> We should have a Makefile with a recipe to install the updates-rules binary. This is also good to have as other parties can install the tool consistently.

👍

**Review by justaugustus — APPROVED** [review]
Echoing from [Slack](https://kubernetes.slack.com/archives/CJH2GBF7Y/p1639423846105100):

@dims @sttts -- Do we feel this is in a place that we could merge/iterate?
I feel like we could have @kubernetes/release-engineering start kicking the tires and cleaning up as needed.

**Comment by justaugustus** [comment]
/assign @nikhita 
ref: [Slack](https://kubernetes.slack.com/archives/CJH2GBF7Y/p1639424089105500?thread_ts=1639423846.105100&cid=CJH2GBF7Y)

**Comment by palnabarun** [comment]
> Question: Should we have alternate yaml loading utils (with yaml.v3) and use it with this tool to retain the comments?
or move the existing utils to yaml.v3?

Parsing YAML comments is still non-trivial and I feel specifically for this feature, the benefits do not outweigh the effort needed.

Looking back at the usage of comments in the latest `rules.yaml`:
```
# TODO: remove bazel related files after we stop publishing branches with
# bazel files
# See: https://github.com/kubernetes/enhancements/issues/2420
```
The above blob of comments can be in a GH issue instead of residing in a comment.

With that, we may not need to parse comments at all and comment parsing becomes a non-blocker.

Thanks @navidshaikh for working on this! ❤️ 

/lgtm

**Comment by navidshaikh** [comment]
> The above blob of comments can be in a GH issue instead of residing in a comment.

👍  
There is only one comment in the rules file related to the [KEP-2420](https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/2420-reducing-kubernetes-build-maintenance/): Reducing Kubernetes Build Maintenance. We can track updating the [rules](https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/rules.yaml) file part of the KEP [tracking issue](https://github.com/kubernetes/enhancements/issues/2420).

cc: @BenTheElder WDYT?

**Comment by justaugustus** [comment]
@kubernetes/publishing-bot-maintainers -- Where did we land with this one?

**Comment by dims** [comment]
/hold cancel
/approve

@palnabarun @justaugustus what are the follow ups after this merges?

**Pull request merged**
